src/net/protocol_agnostic/addr.rs

```diff
@@ -1,6 +1,91 @@
 use priv_prelude::*;
+use url::{self, Url};
+use std::fmt;
+use std::str::FromStr;
+use std::net;
 
 pub enum PaAddr {
     Tcp(SocketAddr),
 }
 
+impl fmt::Display for PaAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            PaAddr::Tcp(ref addr) => {
+                write!(f, "tcp://{}", addr)
+            },
+        }
+    }
+}
+
+quick_error! {
+    #[derive(Debug)]
+    pub enum ParseError {
+        MalformedUrl(e: url::ParseError) {
+            description("malformed url")
+            display("malformed url. {}", e)
+            cause(e)
+        }
+        MalformedHost(e: net::AddrParseError) {
+            description("malformed host")
+            display("malformed host. {}", e)
+            cause(e)
+        }
+        UnknownScheme(s: String) {
+            description("unknown scheme")
+            display("unknown scheme: \"{}\"", s)
+        }
+        MissingPort {
+            description("missing port number")
+        }
+    }
+}
+
+impl FromStr for PaAddr {
+    type Err = ParseError;
+
+    fn from_str(s: &str) -> Result<PaAddr, ParseError> {
+        let url = match Url::parse(s) {
+            Ok(url) => url,
+            Err(e) => return Err(ParseError::MalformedUrl(e)),
+        };
+        let ret = match url.scheme() {
+            "tcp" => {
+                let ip = match url.host_str() {
```

Comment:
It would be nice to move the body of this arm to a separate function: it would be easier to test and `from_str` would be more readable and extendible. E.g. 
```
let ret = match url.scheme() {
    "tcp" => parse_tcp_addr(url),
    "udp" => parse_udp_addr(url),
    _ => Err(...),
}
```